### PR TITLE
feat(core): Allow user role modification

### PR DESCRIPTION
**Pull request by ivov**
https://linear.app/n8n/issue/PAY-985

```
PATCH /users/:id/role
  unauthenticated user
    ✓ should receive 401 (349 ms)
  member
    ✓ should fail to demote owner to member (349 ms)
    ✓ should fail to demote owner to admin (359 ms)
    ✓ should fail to demote admin to member (381 ms)
    ✓ should fail to promote other member to owner (353 ms)
    ✓ should fail to promote other member to admin (377 ms)
    ✓ should fail to promote self to admin (354 ms)
    ✓ should fail to promote self to owner (371 ms)
  admin
    ✓ should receive 400 on invalid payload (351 ms)
    ✓ should receive 404 on unknown target user (351 ms)
    ✓ should fail to demote owner to admin (349 ms)
    ✓ should fail to demote owner to member (347 ms)
    ✓ should fail to promote member to owner (384 ms)
    ✓ should fail to promote admin to owner (350 ms)
    ✓ should be able to demote admin to member (354 ms)
    ✓ should be able to demote self to member (350 ms)
    ✓ should be able to promote member to admin (349 ms)
  owner
    ✓ should be able to promote member to admin (349 ms)
    ✓ should be able to demote admin to member (349 ms)
    ✓ should fail to demote self to admin (348 ms)
    ✓ should fail to demote self to member (354 ms)
```

### CODE REVIEW

File: packages/cli/test/integration/users.api.test.ts
Comment (ivov) [inline packages/cli/test/integration/users.api.test.ts:1]: Tests in this controller are significantly slow. I'll look into this separately.

File: packages/cli/src/requests.ts [hunk 1]
```diff
@@ -322,6 +322,13 @@ export declare namespace UserRequest {
 		{ transferId?: string; includeRole: boolean }
 	>;
 
+	export type ChangeRole = AuthenticatedRequest<
+		{ id: string },
+		{},
+		{ newRole?: { scope?: RoleScopes; name?: RoleNames } },
```
Comment (cstuncsik) [inline packages/cli/src/requests.ts:328]: Maybe I'm just missing something but if it's `ChangeRole` request what would it do without the `newRole` object?
Why is it optional?

Comment (cstuncsik) [reply packages/cli/src/requests.ts:328]: Or is it just to keep the contract real-life like? Like I can send any bullshit request

**Review by cstuncsik — APPROVED** [review]
Just had a question that maybe answered to myself :)
Otherwise it looks good to me

File: packages/cli/src/requests.ts (hunk 1 above)
Comment (ivov) [reply packages/cli/src/requests.ts:328]: > Or is it just to keep the contract real-life like?

Exactly! If we don't do this, when working in the controller we'll assume (based on the typing) that this property is guaranteed to be there, which is not the case and requires validation.

**Pull request merged**